### PR TITLE
Quota limits

### PR DESCRIPTION
**Pull request by iorveth**
Closes https://github.com/Joystream/joystream/issues/2072

### CODE REVIEW

File: runtime-modules/storage/src/data_directory.rs [hunk 1]
```diff
@@ -164,6 +208,11 @@ decl_storage! {
         /// Maps data objects by their content id.
         pub DataObjectByContentId get(fn data_object_by_content_id) config():
             map hasher(blake2_128_concat) T::ContentId => Option<DataObject<T>>;
+
+        pub Quotas get(fn quotas) config():
```
Comment (bedeho) [inline runtime-modules/storage/src/data_directory.rs:252]: Can you add a comment here which describes when a quota is created?

File: runtime-modules/storage/src/data_directory.rs [hunk 2]
```diff
@@ -153,6 +162,41 @@ pub struct DataObjectInternal<
     pub ipfs_content_id: Vec<u8>,
 }
 
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Clone, Encode, Decode, PartialEq, Eq, Debug, Default)]
+pub struct Quota {
```
Comment (bedeho) [inline runtime-modules/storage/src/data_directory.rs:186]: I should have been more clear here that the limit is in terms of size, not in terms of number of objects sorry.

However, I think it actually could make sense to have both, so lets have both kinds of limits+usage in this struct: total size and number of objects.

File: runtime-modules/storage/src/data_directory.rs [hunk 3]
```diff
@@ -342,7 +427,28 @@ decl_module! {
 }
 
 impl<T: Trait> Module<T> {
+    // Ensures quota limit constraint satisfied, returns delta.
+    fn ensure_quota_limit_constraint_satisfied(
```
Comment (bedeho) [inline runtime-modules/storage/src/data_directory.rs]: Very elegant :)

**Review by bedeho — CHANGES_REQUESTED** [review]
Looks like a good start.

- I think we probably also want two global limits, because people can make many channels or memberships and spam. So lets have a global size limit and global object counter limit, which goes across all all owners. These limits should also be adjustable in storage through proposals as before.
- Can we also introduce a blocker variable which blocks all new uploads, sort of like a an emergency switch, can be updated by proposal or lead.

File: runtime-modules/storage/src/data_directory.rs (hunk 2 above)
Comment (iorveth) [reply runtime-modules/storage/src/data_directory.rs:186]: Done here https://github.com/Joystream/joystream/pull/2122/commits/ff989c45bcd0b849082ec16ae25b9adbab5cd13a#diff-5f5bdbc581be08dfecff6870d8fcf9b767904a351d254a70cf35267e5465589bR185

File: runtime-modules/storage/src/data_directory.rs (hunk 1 above)
Comment (iorveth) [reply runtime-modules/storage/src/data_directory.rs:252]: Done https://github.com/Joystream/joystream/pull/2122/commits/ff989c45bcd0b849082ec16ae25b9adbab5cd13a#diff-5f5bdbc581be08dfecff6870d8fcf9b767904a351d254a70cf35267e5465589bR241

File: runtime-modules/storage/src/data_directory.rs (hunk 3 above)
Comment (iorveth) [reply runtime-modules/storage/src/data_directory.rs]: Thank you, it was hard to come up with :) https://github.com/Joystream/joystream/pull/2122/commits/ff989c45bcd0b849082ec16ae25b9adbab5cd13a#diff-5f5bdbc581be08dfecff6870d8fcf9b767904a351d254a70cf35267e5465589bR526

**Comment by iorveth** [comment]
```
I think we probably also want two global limits, because people can make many channels or memberships and spam. So lets have a global size limit and global object counter limit, which goes across all all owners. 
These limits should also be adjustable in storage through proposals as before.
```
Great idea, implemented this here https://github.com/Joystream/joystream/pull/2122/commits/ff989c45bcd0b849082ec16ae25b9adbab5cd13a#diff-5f5bdbc581be08dfecff6870d8fcf9b767904a351d254a70cf35267e5465589bR252

**Comment by iorveth** [comment]
`Can we also introduce a blocker variable which blocks all new uploads, sort of like a an emergency switch, can be updated by proposal or lead.`

Sure, though i feel like it`s better to create a separate issue for that feature.

**Comment by bedeho** [comment]
> Sure, though i feel like it`s better to create a separate issue for that feature.

Agreed.

File: runtime-modules/storage/src/data_directory.rs [hunk 4]
```diff
@@ -162,38 +174,55 @@ pub struct DataObjectInternal<
     pub ipfs_content_id: Vec<u8>,
 }
 
+#[derive(Clone, Copy)]
+pub struct Delta {
```
Comment (bedeho) [inline runtime-modules/storage/src/data_directory.rs]: Can we change this name to `Voucher`, I think quota is not actually fully accurate.

**Review by bedeho — CHANGES_REQUESTED** [review]
Nice fixes, one last improvement, sorry for not pointing out sooner.

File: runtime-modules/storage/src/data_directory.rs (hunk 4 above)
Comment (iorveth) [reply runtime-modules/storage/src/data_directory.rs]: https://github.com/Joystream/joystream/pull/2122/commits/66937035d6a16bbb5985e533c1064139fb8b41c9

**Review by bedeho — APPROVED** [review]
Excellent.

**Pull request merged**